modules/canbus/vehicle/ge3/ge3_controller.cc

```diff
@@ -0,0 +1,838 @@
+      }
+    }
+  }
+
+  // give engage_advice based on error_code and canbus feedback
+  if (chassis_error_mask_) {
+    if (chassis_.has_engage_advice()) {
+      chassis_.mutable_engage_advice()->set_advice(
+          apollo::common::EngageAdvice::DISALLOW_ENGAGE);
+      chassis_.mutable_engage_advice()->set_reason("Chassis error!");
+    }
+  } else if (chassis_.parking_brake() || CheckSafetyError(chassis_detail)) {
+    if (chassis_.has_engage_advice()) {
+      chassis_.mutable_engage_advice()->set_advice(
+          apollo::common::EngageAdvice::DISALLOW_ENGAGE);
+      chassis_.mutable_engage_advice()->set_reason(
+          "Vehicle is not in a safe state to engage!");
+    }
+  } else {
+    if (chassis_.has_engage_advice()) {
+      chassis_.mutable_engage_advice()->set_advice(
+          apollo::common::EngageAdvice::READY_TO_ENGAGE);
+    }
+  }
+
+  return chassis_;
+}
+
+void Ge3Controller::Emergency() {
+  set_driving_mode(Chassis::EMERGENCY_MODE);
+  ResetProtocol();
+  // In emergency case, the hazard lamp should be on
+  pc_bcm_201_->set_pc_hazardlampreq(Pc_bcm_201::PC_HAZARDLAMPREQ_REQ);
+  set_chassis_error_code(Chassis::CHASSIS_ERROR);
+}
+
+ErrorCode Ge3Controller::EnableAutoMode() {
+  if (driving_mode() == Chassis::COMPLETE_AUTO_DRIVE) {
+    AINFO << "already in COMPLETE_AUTO_DRIVE mode";
+    return ErrorCode::OK;
+  }
+  pc_bcs_202_->set_pc_brkpedenable(Pc_bcs_202::PC_BRKPEDENABLE_ENABLE);
+  pc_vcu_205_->set_pc_accpedenable(Pc_vcu_205::PC_ACCPEDENABLE_ENABLE);
+  pc_vcu_205_->set_pc_gearenable(Pc_vcu_205::PC_GEARENABLE_ENABLE);
+  pc_epb_203_->set_pc_epbenable(Pc_epb_203::PC_EPBENABLE_ENABLE);
+  pc_eps_204_->set_pc_steerenable(Pc_eps_204::PC_STEERENABLE_ENABLE);
+
+  can_sender_->Update();
+  const int32_t flag =
+      CHECK_RESPONSE_STEER_UNIT_FLAG | CHECK_RESPONSE_SPEED_UNIT_FLAG;
+  if (!CheckResponse(flag, true)) {
+    AERROR << "Failed to switch to COMPLETE_AUTO_DRIVE mode.";
+    Emergency();
+    set_chassis_error_code(Chassis::CHASSIS_ERROR);
+    return ErrorCode::CANBUS_ERROR;
+  } else {
+    set_driving_mode(Chassis::COMPLETE_AUTO_DRIVE);
+    // if the auto mode can be set normally, the harzad lamp should be off.
+    pc_bcm_201_->set_pc_hazardlampreq(Pc_bcm_201::PC_HAZARDLAMPREQ_NOREQ);
+    AINFO << "Switch to COMPLETE_AUTO_DRIVE mode ok.";
+    return ErrorCode::OK;
+  }
+}
+
+ErrorCode Ge3Controller::DisableAutoMode() {
+  ResetProtocol();
+  can_sender_->Update();
+  set_driving_mode(Chassis::COMPLETE_MANUAL);
+  set_chassis_error_code(Chassis::NO_ERROR);
+  AINFO << "Switch to COMPLETE_MANUAL OK.";
+  return ErrorCode::OK;
+}
+
+ErrorCode Ge3Controller::EnableSteeringOnlyMode() {
+  if (driving_mode() == Chassis::COMPLETE_AUTO_DRIVE ||
+      driving_mode() == Chassis::AUTO_STEER_ONLY) {
+    set_driving_mode(Chassis::AUTO_STEER_ONLY);
+    AINFO << "Already in AUTO_STEER_ONLY mode";
+    return ErrorCode::OK;
+  }
+  pc_bcs_202_->set_pc_brkpedenable(Pc_bcs_202::PC_BRKPEDENABLE_DISABLE);
```

Comment:
same as @mickeyouyou 's comment: for steering only mode, we should not enable brake nor acceleration nor epb.